[dnm] kvserver: investigate memory blowup under testrace

tbg opened the pull request:
I haven't actually been able to reproduce the clock allocator failure. I
am running

```
go test -race -c ./pkg/kv/kvserver/ && export GOMAXPROCS=4 && ./kvserver.test -test.v -test.run TestMergeQueueSeesNonVoters -test.count 30
```

on top of this patch and hitting OOMs (have 32GiB of memory, usually
~16GiB unused).

Pushing this branch so that I can get others to try.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61342)
<!-- Reviewable:end -->

tbg committed:
testcluster: avoid leaking stopper for unstarted cluster

Release justification: non-production code changes
Release note: None

tbg committed:
Updated repro for pebble mem leak

```
rm ./pkg/kv/kvserver/repro/heap.pprof
go test -race ./pkg/kv/kvserver/repro -v
go tool pprof -top ./pkg/kv/kvserver/repro/heap.pprof | grep Showing
Showing nodes accounting for 456.45MB, 95.95% of 475.72MB total
```

When I comment out the finalizer in `pebble.Open` this drops down
to ~7mb.

tbg committed:
<pkg>: <short description - lowercase, no final period>

```
// With finalizer
Showing nodes accounting for 596.71MB, 87.76% of 679.97MB total
// Without finalizer
Showing nodes accounting for 10690.84kB, 100% of 10690.84kB total
```

Release justification:
Release note (<category, see below>): <what> <show> <why>

tbg commented:
@petermattis I still don't understand how that finalizer plays a role, but look at the last commit here - if I comment out the pebble finalizer, our race builds stop leaking memory as far as I can tell.

tbg commented:
I did another round of verifying that the pebble finalizer really is responsible for the rapidly rising memory consumption. With the finalizer left in, looping `TestMergeQueueSeesNonVoters` (which is a beefy 7-node test) blows past 8GB of ram within a minute (and keeps going until oom). With the finalizer commented out, RAM usage stabilizes at ~2.3GiB after around a minute. I think that is pretty conclusive.
